RFC/src/RFC-0173_Versioning.md

```diff
@@ -0,0 +1,94 @@
+shown here.
+
+## Disclaimer
+
+This document and its content are intended for information purposes only and may be subject to change or update
+without notice.
+
+This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
+community. The release of this document is intended solely for review and discussion by the community of the
+technological merits of the potential system outlined herein.
+
+## Goals
+
+The aim of this Request for Comment (RFC) is to describe the various types of versioning that nodes on the Tari network will use while interaction with other nodes.
+
+## Related Requests for Comment
+
+- [RFC-0710: Tari Communication Network and Network Communication Protocol](RFC-0170_NetworkCommunicationProtocol.md)
+- [RFC-0171: MessageSerialization](RFC-0171_MessageSerialisation.md)
+
+## Description
+
+In a decentralized system the set of nodes on the network will run a variety of software versions as time goes on. Some of these versions will be compatible and others not. For example,
+if a crucial consensus change is added during a hard fork event. Furthermore, there will be multiple networks running Tari code, i.e. Mainnet vs Testnet. Versioning refers to the strategies 
+we will use for nodes to determine if they can communicate.
+
+Tari will contain three different versioning schemes:
+1. A WireMode byte that precede every low level network packet,
+2. P2P message versions that will accompany every P2P message,
+3. Consensus rules versions that will be exchanged on connection and are included in each block header.
+
+### WireMode byte
+In the Bitcoin P2P protocol messages are preceded by 4 [magic values](https://en.bitcoin.it/wiki/Protocol_documentation#Common_structures) or bytes. These values are used to 
+delimit when a new message starts in a byte stream and also are used to indicate which of the Bitcoin networks the node is speaking on, such as TestNet or Mainnet.
+
+Tari message packets are encapsulated using the Noise protocol so we do not need the delimiting functionality of these bytes but Tari will include a single WireMode byte at the beginning
+of every connection session. This byte will indicate which network a node is communicating on, so that if the counter-party is on a different network it can reject this connection cheaply without
+having to perform any further operations, like completing the Noise protocol handshake.
+
+### P2P message version
+Peer to Peer messages on the Tari network are encapsulated into message envelopes. The body of message envelopes are defined, serialized and deserialized using Protobuf. These messages will only be updated by adding 
+new fields to the Protobuf definitions, never removing fields. This is done in order to preserve backwards compatibility where newer nodes can still communicate with older nodes, but the reverse
```

Comment:
I may be wrong here, but as far as I have it you can delete Protobuf fields. You just should not reuse the field numbers.

Reply:
Ja, I think you are right.

Reply:
Though by removing the field you make Protobuf not backwards compatible. Should these messages keep the old fields to be able to speak to older nodes?

Reply:
Yeah as I understand it you want to keep them for backwards compatibility?

Reply:
Here's a blog on the topic:
https://www.beautifulcode.co/blog/88-backward-and-forward-compatibility-protobuf-versioning-serialization

We could use explicit version numbers; or handle missing fields in code, which would be a strong indicator of difference versions (but not a guarantee of course). 

In the former case, do you propose integer versioning, or semver-type? If you make a small change to a message that is 100% backwards compatible (wherein the new field will have the default value for the type) and doesn't require any special handling, do you inc the version? Strictly speaking, one should, right? But it feels like overkill.

On the otherhand semver-type versioning feels like a lot of overhead here.

Maybe the compromise is major.minor, where minor is inc with _any_ change, and major with any breaking change.
Then if peer A receives a msg from B where A.version.major > B/major.version, it could (possibly) drop down to the older version if it can, or esle reject the message.
Alternatively, if B get a message from A, it must reject.

The minor version indicator is simply there to mark that the messages aren't 100% identical, but doesn't result in message rejection.

My feeling is that trying to manage versions by dealing with missing info as in the blog post is fraught with unforeseen circumstances and is best avoided.

Reply:
The [Language Guide[(https://developers.google.com/protocol-buffers/docs/proto3#updating) recommends renaming them to "OBSOLETE_" instead of deleting.

I like the idea of having a version number in the message header. Since missing fields are rendered as defaults in protobuf 3, nodes can reject newer versions, and handle missing fields in older versions. I think handling defaults in the protobuf conversions is tricky even at the best of times, so care will need to be taken in these parts of code. I also think the changes on the p2p messages will be small enough that semver is not needed and there is a good chance that the minor versions will not be used

Reply:
Wrote up a two-part semantic versioning proposal for the P2P messages